test/unit-test/mqtt_agent_utest.c

```diff
@@ -74,21 +110,455 @@ int suiteTearDown( int numFailures )
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.getCommand = mockGetCommand;
+    agentContext.agentInterface.releaseCommand = NULL;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.releaseCommand = mockReleaseCommand;
+    agentContext.agentInterface.pMsgCtx = NULL;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+
+    agentContext.agentInterface.pMsgCtx = &globalMessageContext;
+    /* Invalid packet ID to indicate uninitialized context. */
+    agentContext.mqttContext.nextPacketId = MQTT_PACKET_ID_INVALID;
+    mqttStatus = MQTTAgent_Terminate( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTBadParameter, mqttStatus );
+}
+
+/**
+ * @brief Test error cases when a command cannot be obtained or sent
+ * for API functions that create commands.
+ */
+void test_MQTTAgent_API_Create_Command_Creation_Error( void )
+{
+    MQTTAgentContext_t agentContext = { 0 };
+    MQTTStatus_t mqttStatus;
+    CommandInfo_t commandInfo = { 0 };
+    Command_t command = { 0 };
+
+    setupAgentContext( &agentContext );
+
+    /* Use Ping since it's one that doesn't have an additional pointer param. */
+    pCommandToReturn = NULL;
+    mqttStatus = MQTTAgent_Ping( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTNoMemory, mqttStatus );
+
+    pCommandToReturn = &command;
+    agentContext.agentInterface.send = mockSendFail;
+    mqttStatus = MQTTAgent_Ping( &agentContext, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTSendFailed, mqttStatus );
+    /* Test that the command was released. */
+    TEST_ASSERT_EQUAL_INT( 1, commandReleaseCallCount );
+    /* Also test that the command was set. */
+    TEST_ASSERT_EQUAL( PING, command.commandType );
+}
+
+/**
+ * @brief Test that a bad parameter is returned when there
+ * is no more space to store a pending acknowledgment for
+ * a command that expects one.
+ */
+void test_MQTTAgent_API_Create_Command_No_Ack_Space( void )
+{
+    MQTTAgentContext_t agentContext;
+    MQTTStatus_t mqttStatus;
+    CommandInfo_t commandInfo = { 0 };
+    Command_t command = { 0 };
+    MQTTSubscribeInfo_t subscribeInfo = { 0 };
+    MQTTAgentSubscribeArgs_t subscribeArgs = { 0 };
+    int i;
+
+    setupAgentContext( &agentContext );
+    pCommandToReturn = &command;
+
+    /* No space in pending ack list. */
+    for( i = 0; i < MQTT_AGENT_MAX_OUTSTANDING_ACKS; i++ )
+    {
+        agentContext.pPendingAcks[ i ].packetId = ( i + 1 );
+    }
+
+    subscribeArgs.pSubscribeInfo = &subscribeInfo;
+    subscribeArgs.numSubscriptions = 1U;
+    mqttStatus = MQTTAgent_Subscribe( &agentContext, &subscribeArgs, &commandInfo );
+    TEST_ASSERT_EQUAL( MQTTNoMemory, mqttStatus );
+}
+
+/**
+ * @brief Test that MQTTAgent_Subscribe() works as intended.
+ */
+void test_MQTTAgent_Subscribe( void )
```

Comment:
Can be broken in three separate tests - NULL params, incorrect params, happy path.